src/ng/directive/input.js

```diff
@@ -1431,17 +1431,20 @@ function createDateInputType(type, regexp, parseDate, format) {
   return function dynamicDateInputType(scope, element, attr, ctrl, $sniffer, $browser, $filter) {
     badInputChecker(scope, element, attr, ctrl);
     baseInputType(scope, element, attr, ctrl, $sniffer, $browser);
-    var timezone = ctrl && ctrl.$options.getOption('timezone');
     var previousDate;
 
     ctrl.$$parserName = type;
     ctrl.$parsers.push(function(value) {
-      if (ctrl.$isEmpty(value)) return null;
+      if (ctrl.$isEmpty(value)) {
+        previousDate = null;
```

Comment:
If it's not valid, then why are there no test failures? :D
The formatter is doing the same thing afaict - it sets previousDate = null if the model is not a valid date (empty is not a valid date, too).
The specific test at https://github.com/angular/angular.js/pull/16336/files/d172e8ecb650ad3220e00e46b08325ad3672e9f1#diff-ce6d0f141f60d25eede7abf044d04b93R1968 shows that when clearing out the input, previousDate is still set and this leads the next input to use an incorrect value for the hour part.

We'd need a test case to show that this is a regression.

Reply:
> If it's not valid, then why are there no test failures? :D

Valid point...nooot 😛 

> The formatter is doing the same thing afaict - it sets `previousDate = null` if the model is not a valid date (empty is not a valid date, too).

That is the point. The formatter processes the model value and `previousDate` keeps track of the last valid (date) model value.

> The specific test at https://github.com/angular/angular.js/pull/16336/files/d172e8ecb650ad3220e00e46b08325ad3672e9f1#diff-ce6d0f141f60d25eede7abf044d04b93R1968 shows that when clearing out the input, previousDate is still set and this leads the next input to use an incorrect value for the hour part.

"use an incorrect value for the hour part" is subjective. I'd say currently it uses "the correct value for the hour part".
The idea is to be able to set the hour for a date field. AFAICT, the only way to achieve this today is by setting the model to a date with the desired date. Then, as the user edits the field (including emptying and refilling it) the hour/minute/second part is preserved. I think this is a desired property (I am pretty sure I've used it in my apps :smiley:).
In any case, I believe this change is breaking not related to this PR. It should go into a separate PR (where it can be debated further :grin:).

Reply:
> The idea is to be able to set the hour for a date field.

Ah, that's right. Did you get this from the code or did you search for the commit? Here's the commit: https://github.com/angular/angular.js/commit/1a1ef62903c8fdf4ceb81277d966a8eff67f0a96

As you can see, there are not tests for type=date ... :<

This is at least confusing, because as you can see in the test, the date is re-using the hour part that was set for -0500, even though the timezone is set to UTC again. But maybe this is a contrived example.

Reply:
I didn't search for the commit (I didn't expect there to be a commit specific for that - I assumed it was part of the initial implementation) :smiley:

So, I am not sure what the correct behavior is here (wrt to changing the timezone) 😕
Maybe keep track of the previous timezone and re-adjust? Or accept the fact that changing the timezone may change the time part?
Timezones are tricky :grin: